File: src/bin/0020-valid-parentheses.cpp
// @leetup=info id=20 lang=cpp slug=valid-parentheses

#include <algorithm>
#include <cassert>
#include <iostream>
#include <map>
#include <set>
#include <stack>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  bool isValid(std::string s) {
    std::stack<char> stack;
    std::map<char, char> pairs{{')', '('}, {'}', '{'}, {']', '['}};

    for (char c : s) {
      auto it = pairs.find(c);
      if (it != pairs.end()) {
        if (stack.empty()) {
          return false;
        }
        char top = stack.top();
        stack.pop();
        if (it->second != top) {
          return false;
        }
      } else {
        stack.push(c);
      }
    }
    return stack.empty();
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0020-valid-parentheses.py
# @leetup=info id=20 lang=python3 slug=valid-parentheses



from typing import Optional

# @leetup=code
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {
            ")": "(",
            "}": "{",
            "]": "[",
        }
        for a in s:
            if a in pairs:
                if not stack or pairs[a] != stack.pop():
                    return False
            else:
                stack.append(a)
        return len(stack) == 0


# @leetup=code
